### ISSUE TITLE
PSDReader loads wrong band indicies in grayscale psd files with backgrounds.

### ISSUE THREAD

**Issue by Eurybiadan** [haraldk/TwelveMonkeys#248]
The band index creation section (PSDImageReader: line 1116) is this:

```
for (int i = 0, indicesLength = indices.length; i < indicesLength; i++) {
     indices[i] = indicesLength - i;
}

```
And when a layer from grayscale psd with a background is loaded, it fills in the index array as follows:

```
indicies[0] = 2;
indicies[1] = 1;
```

Which is incorrect, because the bands (in this case) should be loaded sequentially (0,1).

**Comment by Eurybiadan** [comment]
I've attached a JUnit test case for this issue. It should be easy to integrate into the existing JUnit tests in its current state, if desired... otherwise I can issue a pull request with the fix and the test.

[Bandreading_JUnit.zip](https://github.com/haraldk/TwelveMonkeys/files/273636/Bandreading_JUnit.zip)

**Comment by haraldk** [comment]
Thanks,

Unfortunately, I can't use the code as it contains a GNU license, which is incompatible with the BSD license used in the project. Feel free to resubmit without it, or with a compatible (BSD) license.

Also note that simply testing that the band indices are in order is too strict. Common configurations for RGB + alpha, are ARGB (3, 0, 1, 2), ABGR (3, 2, 1, 0), BGRA (2, 1, 0, 3) or RGBA (0, 1, 2, 3).

We should probably instead see that the colors/alpha values in the respective bands are correct. 


Regards,

-- 
Harald K

**Comment by Eurybiadan** [comment]
Sorry about the delay @haraldk - other projects have consumed my time.

I apologize for the license issues, they'll be fixed when I sort this out.

I'm running in small circles right now; I can't seem to find a test case that will test the band indices correctly. For example, if I leave the code as-is and load a rgb PSD with alpha, the indices are 4,3,2,1. This appears to correctly load the psd, and the layers are the correct colors.

However, if I use the opposite indices (0,1,2,3), the psd is again correctly loaded, and the layers are again the correct colors (I would expect inverted colors).

I'm still trying to pull this apart, but ideas are welcome...

**Comment by Eurybiadan** [comment]
So it looks like the RGB psd loading is "rescued" by getImageTypes (PSDImageReader line 259), where a pair of standard interleaved ImageTypeSpecifiers are created, in addition to the banded specifier (which has been the primary focus of my efforts). After it creates the specifiers, it creates the destination image. But instead of using the banded specifier, it uses the first interleaved specifier (ABGR, which is indices 3,2,1,0) as a fallback (ImageReaderBase line 284).

Perhaps I will need to develop a test case that bypasses the added interleaved specifiers. More likely, we will need a special case for grayscale.

**Comment by Eurybiadan** [comment]
I have test data that I can do that with; I'll name the layers according to their color content (or grayscale value) as reported by Photoshop, then we can do a "getRGB"-derived check against that. I know getRGB is slow, but it's a test case, not production code.

Does that seem reasonable?

**Comment by haraldk** [comment]
That would be great!

If you create the sample file, I can create the test case.
Would be nice with:
* one RGB file, with a couple (2-3) layers with alpha and same colors without alpha (4-6 layers total)
* one gray file with 2-3 layers with alpha and same without alpha

(If mixing alpha and non-alpha layers is hard, we could also use 4 sample files).

The smaller the sample files, the better. :-)


Thanks,

-- 
Harald K

**Comment by Eurybiadan** [comment]
Done.

However, yesterday I put together a grayscale test case on my end (just to make sure everything worked properly), the values seem to be read incorrectly. Specifically, the DataBuffer contains gamma-corrected values, instead of the "true" values.

For example:

(Actual grayscale value / DataBuffer value)
59/42
89/68
100/78
115/92
150/129
200/186
255/255

If you plot these, it looks like a gamma correction. I'm not sure why this is. I've debugged it all the way down to the byte-read (PSDImageReader line 660), and the value is being read in pre-shifted.

Ideas are welcome, I'm kind of scratching my head over here...

**Comment by haraldk** [comment]
I'm a little confused about what I am looking at here... What are the "Actual grayscale value"s? The ones you have entered in Photoshop? Or the ones you see on your screen? Or the ones reported by `bi.getRGB(...)`?

Could be caused by the `convertToDestinationCS` method, that converts the data from the source to destination color space, if these are different. But I don't know that, as I don't have your sample files. :-)

Anyway, looking at the embedded profile of the `test_grayscale_boxes.psd` file that I do have (Adobe, "Dot Gain 20%") and the one used in Java by default for grayscale (Kodak, "KODAK Grayscale Conversion - Gamma 1.0"), this might make sense (the built in gray is from my Oracle Java 1.8u2 build, yours may differ). 

In any case, the color management used in Photoshop and in Java (either Kodak or LCMS) are different. So, it is to be expected that the colors also varies slightly (although perhaps not this much, but again, I don't know what I am looking at).


-- 
Harald K

**Comment by Eurybiadan** [comment]
Sorry for the confusion. The "actual" grayscale values are the ones entered/assigned in Photoshop.

I'm using a modified version of test_grayscale_boxes.psd (attached, you'll need to rename), that I still need to shrink down a bit for the formal test case.

![test_grayscale_boxes](https://cloud.githubusercontent.com/assets/5322065/15898384/203f15a6-2d66-11e6-90c3-074d40517485.png)

What I find confusing is that the values are incorrect at readFully, when they haven't even been put into the DataBuffer yet. I suppose it is possible that Photoshop tweaks the values at save, and converts them, but re-saving the file without the embedded profile doesn't change the values.

From what I can tell, `convertToDestinationCS` is never called.

**Comment by haraldk** [comment]
Ok,

I'm pretty sure what you see is completely normal then. :-)

As the PSD is stored with the "Dot Gain 20%" profile, the values in the file will not match the values in Photoshop. Thus, the values in the `DataBuffer` will also be different. This is not incorrect. Just different. :-)

However, if you invoke `image.getRGB(x, y) & 0xFF` (the R, G and B values should be pretty much the same here), you should get about the same value back (apart from the mentioned differences from the different CMSes). Using a different profile should either write different values or display the image differently in Photoshop. If you save without a profile, I don't know...

PS: I have a modified version of the code here, where I add some extra types to the type specifiers, that is why `convertToDestinationCS` is invoked (for the gray with no alpha) for me. Never mind that. :-)

-- 
Harald K

**Comment by Eurybiadan** [comment]
> If you save without a profile, I don't know...

This is exactly what's bugging me :-/

**Comment by Eurybiadan** [comment]
Last night I confirmed that changing the profiles in Photoshop changes the appearance of the layers in Java. I switched between a Dot Gain 20%, and a Gamma 1.1, and they looked distinctly different.

I'm using this library for scientific programming, and that makes my work sensitive to the "true" values of the layers. It appears that I'll need to write a colorconvertop to convert the values back, then...
